src/Microsoft.Unity.Analyzers/ContextMenuSuppressor.cs

```diff
@@ -0,0 +1,126 @@
+
+		public override void ReportSuppressions(SuppressionAnalysisContext context)
+		{
+			foreach (var diagnostic in context.ReportedDiagnostics)
+			{
+				AnalyzeDiagnostic(diagnostic, context);
+			}
+		}
+
+		public override ImmutableArray<SuppressionDescriptor> SupportedSuppressions =>
+			ImmutableArray.Create(ContextMenuRule, ContextMenuItemUnusedRule, ContextMenuItemReadonlyRule);
+
+		private void AnalyzeDiagnostic(Diagnostic diagnostic, SuppressionAnalysisContext context)
+		{
+			var sourceTree = diagnostic.Location.SourceTree;
+			var root = sourceTree.GetRoot(context.CancellationToken);
+			var node = root.FindNode(diagnostic.Location.SourceSpan);
+			var model = context.GetSemanticModel(diagnostic.Location.SourceTree);
+			var symbols = new List<ISymbol>();
+
+			switch (node)
+			{
+				case MethodDeclarationSyntax method:
+					symbols.Add(model.GetDeclaredSymbol(method));
+					break;
+				case VariableDeclaratorSyntax vdec:
+					symbols.Add(model.GetDeclaredSymbol(vdec));
+					break;
+				case FieldDeclarationSyntax fdec:
+					symbols.AddRange(fdec.Declaration.Variables.Select(v => model.GetDeclaredSymbol(v)));
+					break;
+			}
+
+			var reportableSymbols = symbols
+				.Where(IsReportable)
+				.ToList();
+
+			if (!reportableSymbols.Any())
+				return;
+
+			if (reportableSymbols.FirstOrDefault() is IMethodSymbol)
+				context.ReportSuppression(Suppression.Create(ContextMenuRule, diagnostic));
+			else
+				foreach (var descriptor in SupportedSuppressions.Where(d => d.SuppressedDiagnosticId == diagnostic.Id))
+					context.ReportSuppression(Suppression.Create(descriptor, diagnostic));
+		}
+
+		private static bool IsReportable(ISymbol symbol)
+		{
+			if (symbol == null)
+				return false;
+
+			var containingType = symbol.ContainingType;
+			if (!containingType.Extends(typeof(UnityEngine.MonoBehaviour)))
+				return false;
+
+			switch (symbol)
+			{
+				case IMethodSymbol methodSymbol:
+					if (methodSymbol.GetAttributes().Any(a => a.AttributeClass.Matches(typeof(UnityEngine.ContextMenu))))
+						return true;
+					if (MethodIsContextMenuCommand(methodSymbol, containingType))
+						return true;
+					break;
+				case IFieldSymbol fieldSymbol:
+					if (fieldSymbol.GetAttributes().Any(a => a.AttributeClass.Matches(typeof(UnityEngine.ContextMenuItemAttribute))))
+						return true;
+					break;
+			}
+
+			return false;
+		}
+
+		private static bool MethodIsContextMenuCommand(IMethodSymbol symbol, INamedTypeSymbol containingType)
+		{
+			foreach (var member in containingType.GetMembers())
+			{
+				if (!(member is IFieldSymbol fieldSymbol))
+					continue;
+
+				var attributes = fieldSymbol.GetAttributes().Where(a => a.AttributeClass.Matches(typeof(UnityEngine.ContextMenuItemAttribute)));
+
+				if (!attributes.Any())
+					continue;
+
+				foreach (var attribute in attributes)
+					if (symbol.Name == attribute.ConstructorArguments[1].Value.ToString())
```

Comment:
@jbevain Is there a better way to do this that doesn't rely on a string comparison?

Reply:
That's also a nasty nullref risk there, I think. Fixing now.